File: web/src/identity/wallet.js
// web/src/identity/wallet.js
// Mirrors go/internal/identity/wallet.go: prf -> BIP39 -> SLIP-0010-ed25519
// m/44'/501'/0'/0' -> Ed25519 -> base58 Solana address. Independent of the
// X25519 transport key (owner.js); the two share only the prf root.
import { ed25519 } from '@noble/curves/ed25519';
import { entropyToMnemonic, mnemonicToSeed } from '../wallet/bip39.js';
import { derivePath } from '../wallet/slip10.js';
import { encode as base58encode } from '../wallet/base58.js';

// WALLET_PATH is the Phantom-importable Solana account-0 path.
export const WALLET_PATH = "m/44'/501'/0'/0'";

// deriveWallet renders the 32-byte prf as a BIP39 mnemonic and derives the
// account-0 Solana wallet. Returns { mnemonic, seed, priv, pub, address } with
// priv = the 32-byte Ed25519 node key (the signing key for @noble/curves).
export function deriveWallet(prf) {
  return walletFromMnemonic(entropyToMnemonic(prf));
}

// walletFromMnemonic derives the account-0 wallet from a mnemonic (import path).
export function walletFromMnemonic(mnemonic) {
  const seed = mnemonicToSeed(mnemonic, '');
  const node = derivePath(seed, WALLET_PATH);
  const pub = ed25519.getPublicKey(node.key);
  return {
    mnemonic,
    seed,
    priv: node.key, // 32-byte ed25519 seed/private key
    pub, // 32-byte ed25519 public key
    address: base58encode(pub),
  };
}

File: go/internal/identity/wallet.go
// go/internal/identity/wallet.go
package identity

import (
	"crypto/ed25519"

	"github.com/srcful/terminal-relay/go/internal/base58"
	"github.com/srcful/terminal-relay/go/internal/bip39"
	"github.com/srcful/terminal-relay/go/internal/slip10"
)

// WalletPath is the Phantom-importable Solana account-0 derivation path.
// Sub-accounts use m/44'/501'/i'/0'.
const WalletPath = "m/44'/501'/0'/0'"

// Wallet is the Solana-compatible Ed25519 identity derived from the passkey prf.
// It is independent of the X25519 transport key (DeriveOwnerKey); the two share
// only the prf root. Address (base58 of the public key) is the wallet owner_id.
type Wallet struct {
	Mnemonic string             // 24-word BIP39 rendering of the prf
	Seed     []byte             // 64-byte BIP39 seed
	Priv     ed25519.PrivateKey // 64-byte (seed||pub); Priv.Seed() is the 32-byte node key
	Pub      ed25519.PublicKey  // 32-byte Ed25519 public key
	Address  string             // base58(Pub) — the Solana address / wallet owner_id
}

// DeriveWallet renders the 32-byte prf as a BIP39 mnemonic and derives the
// account-0 Solana wallet. Restoring from the mnemonic reconstructs prf and
// re-derives the same wallet without the passkey. Mirrors
// web/src/identity/wallet.js exactly.
func DeriveWallet(prf []byte) (*Wallet, error) {
	mnemonic, err := bip39.EntropyToMnemonic(prf)
	if err != nil {
		return nil, err
	}
	return WalletFromMnemonic(mnemonic)
}

// WalletFromMnemonic derives the account-0 wallet from a BIP39 mnemonic (the
// import path). Empty BIP39 passphrase, matching DeriveWallet.
func WalletFromMnemonic(mnemonic string) (*Wallet, error) {
	seed := bip39.MnemonicToSeed(mnemonic, "")
	node, err := slip10.DerivePath(seed, WalletPath)
	if err != nil {
		return nil, err
	}
	priv := ed25519.NewKeyFromSeed(node.Key)
	pub := priv.Public().(ed25519.PublicKey)
	return &Wallet{
		Mnemonic: mnemonic,
		Seed:     seed,
		Priv:     priv,
		Pub:      pub,
		Address:  base58.Encode(pub),
	}, nil
}
